src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/LspFileBasedProgramWorkspace.cs

```diff
@@ -0,0 +1,323 @@
+            TimeSpan.FromMilliseconds(100),
+            LoadOrReloadProjectsAsync,
+            equalityComparer: null,
+            listenerProvider.GetListener(FeatureAttribute.Workspace),
+            CancellationToken.None);
+        iLogger = loggerFactory.CreateLogger<LspFileBasedProgramWorkspace>();
+    }
+
+    public async Task<Document?> AddMiscellaneousDocumentAsync(Uri uri, SourceText documentText, string languageId, ILspLogger logger, CancellationToken cancellationToken)
+    {
+        var documentFilePath = ProtocolConversions.GetDocumentFilePathFromUri(uri);
+        var documentId2 = this.CurrentSolution.GetDocumentIdsWithFilePath(documentFilePath);
+        var document2 = this.CurrentSolution.GetDocument(documentId2.FirstOrDefault());
+        if (document2 != null)
+        {
+            return document2;
+        }
+
+        var languageInfoProvider = _lspServices.GetRequiredService<ILanguageInfoProvider>();
+        if (!languageInfoProvider.TryGetLanguageInformation(uri, languageId, out var languageInformation))
+        {
+            // Only log here since throwing here could take down the LSP server.
+            logger.LogError($"Could not find language information for {uri} with absolute path {documentFilePath}");
+            return null;
+        }
+
+        var services = Services.SolutionServices;
+        var languageServices = services.GetLanguageServices(languageInformation.LanguageName);
+        var parseOptions = languageServices.GetService<ISyntaxTreeFactoryService>()?.GetDefaultParseOptionsWithLatestLanguageVersion();
+
+        // It seems like the workflow in main branch is:
+        // 1. add file to misc workspace
+        // 2. process the project in background
+        // 3. realize file is part of the project
+        // 4. remove from misc workspace and reopen in host workspace
+        // It might be expensive if (1) effectively changes from "just open in misc workspace" versus "generate/process it as its own project etc.".
+        // We might need a stronger heuristic that the current file is unlikely to be part of a project. For example, that there is no .csproj in any of the containing directories.
+        if (!checkIsFileBasedProgram())
+        {
+            return null;
+        }
+
+        using (await _gate.DisposableWaitAsync(cancellationToken))
+        {
+            _projectsToLoadAndReload.AddWork((uri, documentText, languageId));
+
+            // Wait for the in progress batch to complete and send a project initialized notification to the client.
+            await _projectsToLoadAndReload.WaitUntilCurrentBatchCompletesAsync();
+            var documentId = this.CurrentSolution.GetDocumentIdsWithFilePath(documentFilePath).First();
+            var document = this.CurrentSolution.GetDocument(documentId);
+            return document;
+        }
+
+        bool checkIsFileBasedProgram()
```

Comment:
IDE local functions are PascalCased.  that way we can distinguish delegate instances (camel) vs functions (pascal).